src/hyperscan.rs

```diff
@@ -0,0 +1,299 @@
+            id: P_IP.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::EMAIL) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_EMAIL).unwrap().clone(),
+            id: P_EMAIL.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::DATE) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_DATE).unwrap().clone(),
+            id: P_DATE.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::QUOTED) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_QUOTED).unwrap().clone(),
+            id: P_QUOTED.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::URL) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_URL).unwrap().clone(),
+            id: P_URL.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::PHONE) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_PHONE).unwrap().clone(),
+            id: P_PHONE.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+    if flags.contains(constants::ScanFlags::USER_AGENT) {
+        patterns.push(Pattern {
+            expression: pattern_list.get(&P_USER_AGENT).unwrap().clone(),
+            id: P_USER_AGENT.clone(),
+            flags: CompileFlags(HS_FLAG_CASELESS | HS_FLAG_SOM_LEFTMOST),
+        });
+    }
+
+    let res_db: Result<BlockDatabase, _> = patterns.build();
+
+    match res_db {
+        Ok(db) => Some(db),
+        Err(_) => None,
+    }
+}
+
+pub struct HSPatternMatch {
+    pub pattern_id: u32,
+    pub from: u64,
+    pub to: u64,
+}
+
+pub struct HSLineScanner<'a> {
+    //    pub inner: Arc<Mutex<LineScannerData>>,
+    pub lines: &'a Vec<String>,
+    pub line_matches: HashMap<u16, HashMap<u16, Vec<HSPatternMatch>>>,
+}
+
+pub type HSPatternMatchResults =
+    Arc<RwLock<HashMap<u16, RwLock<HashMap<u16, Vec<HSPatternMatch>>>>>>;
+
+impl<'a> HSLineScanner<'a> {
+    pub fn new(lines: &Vec<String>) -> HSLineScanner {
+        HSLineScanner {
+            lines: lines,
+            line_matches: HashMap::new(),
+        }
+    }
+
+    pub fn scan(&mut self, db: &mut BlockDatabase) -> HSPatternMatchResults {
+        let now = Instant::now();
+
+        let line_total = self.lines.len();
+        let scratch = db.alloc().unwrap();
+
+        let pattern_match_results: HSPatternMatchResults = Arc::new(RwLock::new(HashMap::new()));
+
+        for i in 0..line_total {
```

Comment:
Scan we scan a full set of lines in a single scan_() call?

This may greatly reduce the setup/initialization costs (be interesting to get an idea of the performance impact).

Reply:
Yes, we scan batches for `4096` lines at a time, we used to do processing line by line, it's easy to return to that size and test